bindings/rust/aws-kms-tls-auth/src/receiver.rs

```diff
@@ -0,0 +1,358 @@
+    /// 3. cache the decrypted material in the key cache
+    /// 4. return an s2n-tls psk
+    ///
+    /// All of the arguments are owned to satisfy the `'static` bound that s2n-tls
+    /// requires on connection futures.
+    async fn kms_decrypt_and_update(
+        psk_identity: Vec<u8>,
+        ciphertext_datakey: Vec<u8>,
+        client: Client,
+        trusted_key_arns: Arc<Vec<KeyArn>>,
+        key_cache: Cache<Vec<u8>, Vec<u8>>,
+    ) -> anyhow::Result<s2n_tls::psk::Psk> {
+        let ciphertext_datakey_clone = ciphertext_datakey.clone();
+        let decrypted = tokio::spawn(async move {
+            client
+                .decrypt()
+                .ciphertext_blob(Blob::new(ciphertext_datakey_clone))
+                .send()
+                .await
+        })
+        .await??;
+
+        // although the field is called `key_id`, it is actually the key arn. This
+        // is confirmed in the documentation:
+        // https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html#API_Decrypt_ResponseSyntax
+        let associated_key_arn = decrypted.key_id.as_ref().unwrap();
+        if !trusted_key_arns.contains(associated_key_arn) {
+            anyhow::bail!("untrusted KMS Key: {associated_key_arn} is not trusted");
+        }
+
+        let plaintext_datakey = decrypted.plaintext.unwrap().into_inner();
+
+        key_cache.insert(ciphertext_datakey, plaintext_datakey.clone());
+
+        let psk = psk_from_material(&psk_identity, &plaintext_datakey).unwrap();
+
+        Ok(psk)
+    }
+}
+
+impl ClientHelloCallback for PskReceiver {
+    fn on_client_hello(
+        &self,
+        connection: &mut s2n_tls::connection::Connection,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, s2n_tls::error::Error> {
+        // parse the identity list from the client hello
+        let client_hello = connection.client_hello()?;
+        let identities = match retrieve_psk_identities(client_hello) {
+            Ok(identities) => identities,
+            Err(e) => {
+                return Err(s2n_tls::error::Error::application(e.into()));
+            }
+        };
+
+        // extract the identity bytes
+        let psk_identity = match identities.list().first() {
+            Some(id) => id.identity.blob(),
+            None => {
+                return Err(s2n_tls::error::Error::application(
+                    "identities list was zero-length".into(),
+                ))
+            }
+        };
+
+        // parse the identity bytes to a KmsTlsPskIdentity
+        let identity = PskIdentity::decode_from_exact(psk_identity)
+            .map_err(|e| s2n_tls::error::Error::application(e.into()))?;
+
+        // deobfuscate the identity to get the ciphertext datakey
+        let ciphertext_datakey = identity
+            .deobfuscate_datakey(&self.obfuscation_keys)
+            .map_err(|e| s2n_tls::error::Error::application(e.into()))?;
+
+        let maybe_cached = self.key_cache.get(&ciphertext_datakey);
```

Comment:
So if I'm understanding the flow right, I can only trigger a call to KMS if I have a valid, obfuscated identity. If I try to just provide gibberish, then this decrypt will fail and we'll never even do the cache lookup.

But I don't necessarily need the obfuscation key to have valid identities. I could capture a bunch of valid identities and then replay them later once they're expired to trigger KMS calls. If I capture enough identities, I can also reuse them immediately because the cache will evict them early if it fills up.